src/prof_data.c

```diff
@@ -1444,9 +1441,20 @@ prof_tctx_destroy(tsd_t *tsd, prof_tctx_t *tctx) {
 	assert(tctx->cnts.accumobjs == 0);
 	assert(tctx->cnts.accumbytes == 0);
 
-	ckh_remove(tsd, &tdata->bt2tctx, &gctx->bt, NULL, NULL);
-	destroy_tdata = prof_tdata_should_destroy(tsd_tsdn(tsd), tdata, false);
-	malloc_mutex_unlock(tsd_tsdn(tsd), tdata->lock);
+	prof_gctx_t *gctx = tctx->gctx;
+
+	{
+		prof_tdata_t *tdata = tctx->tdata;
+		ckh_remove(tsd, &tdata->bt2tctx, &gctx->bt, NULL, NULL);
+		bool destroy_tdata = prof_tdata_should_destroy(tsd_tsdn(tsd),
+		    tdata, false);
+		malloc_mutex_unlock(tsd_tsdn(tsd), tdata->lock);
+		if (destroy_tdata) {
+			prof_tdata_destroy(tsd, tdata, false);
```

Comment:
If we end up destroying tdata here, does that imply tctx is guaranteed to be destroyed by end of this function? If so, let's assert that.

Reply:
No, it's not guaranteed. It's possible that this function ends but `tctx` is not destroyed, if `tctx` is being dumped by another thread, and in such a case the dumping thread will call `prof_gctx_finish()` in the end to destroy `tctx`.

Reply:
Okay in that case, when the delayed destroying tctx happens, does it touch tdata? Either way, let's clear tctx->tdata so that it's easier to catch.